HBASE-23073 Add an optional costFunction to balance regions according to a capacity rule

PierreZ opened the pull request:
Based on the work in HBASE-22618, users can now load custom costFunctions inside the main balancer used by HBase. As an example, we like like to add upstream an optional cost function called HeterogeneousRegionCountCostFunction that will deal with our issue: how to balance regions according to the capacity of a RS instead of using the RegionCountSkewCostFunction that is trying to avoid skew.

A rule file is loaded from HDFS before balancing. It contains lines of rules. A rule is composed of a regexp for hostname, and a limit. For example, we could have:

    rs[0-9] 200

    rs1[0-9] 50

RegionServers with hostname matching the first rules will have a limit of 200, and the others 50. If there's no match, a default is set.

Thanks to the rule, we have two informations: the max number of regions for this cluster, and the rules for each servers. HeterogeneousBalancer will try to balance regions according to their capacity.

Let's take an example. Let's say that we have 20 RS:

10 RS, named through rs0 to rs9 loaded with 60 regions each, and each can handle 200 regions.
10 RS, named through rs10 to rs19 loaded with 60 regions each, and each can support 50 regions.

Based on the following rules:

rs[0-9] 200

rs1[0-9] 50

The second group is overloaded, whereas the first group has plenty of space. Moving a region from the first group to the second should provide a lower cost.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 51s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 31s |  master passed  |
| :green_heart: |  shadedjars  |   4m 56s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 37s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 25s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 23s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 24s |  the patch passed  |
| :green_heart: |  compile  |   1m  0s |  the patch passed  |
| :green_heart: |  javac  |   1m  0s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 28s |  hbase-server: The patch generated 32 new + 86 unchanged - 19 fixed = 118 total (was 105)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :broken_heart: |  shadedjars  |   3m 59s |  patch has 10 errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m  2s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 37s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 52s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  |  30m 27s |  hbase-server in the patch failed.  |
| :broken_heart: |  asflicense  |   0m 15s |  The patch generated 1 ASF License warnings.  |
|  |   |  90m 54s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.balancer.TestStochasticLoadBalancerHeterogeneousCostRules |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 9f6b64936f0e 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 52176184df |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/artifact/out/diff-checkstyle-hbase-server.txt |
| shadedjars | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/artifact/out/patch-shadedjars.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/testReport/ |
| asflicense | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 678 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   6m 45s |  master passed  |
| :green_heart: |  compile  |   0m 57s |  master passed  |
| :green_heart: |  checkstyle  |   1m 33s |  master passed  |
| :green_heart: |  shadedjars  |   4m 54s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 40s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 32s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 30s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 25s |  the patch passed  |
| :green_heart: |  compile  |   0m 55s |  the patch passed  |
| :green_heart: |  javac  |   0m 55s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 26s |  hbase-server: The patch generated 17 new + 86 unchanged - 19 fixed = 103 total (was 105)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 59s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 23s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 43s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  |  30m  0s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 16s |  The patch does not generate ASF License warnings.  |
|  |   |  92m 23s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.balancer.TestStochasticLoadBalancerHeterogeneousCostRules |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux cd5f63a28f70 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 2ebdcbc3b0 |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/2/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/2/testReport/ |
| Max. process+thread count | 669 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

PierreZ commented:
I need to add a few more tests before it can be merged, for example load the rule file from HDFS, and more cases on the balancing itself.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 27s |  master passed  |
| :green_heart: |  compile  |   0m 55s |  master passed  |
| :green_heart: |  checkstyle  |   1m 22s |  master passed  |
| :green_heart: |  shadedjars  |   4m 37s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 39s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 11s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m  9s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   4m 59s |  the patch passed  |
| :green_heart: |  compile  |   0m 55s |  the patch passed  |
| :green_heart: |  javac  |   0m 55s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 20s |  hbase-server: The patch generated 17 new + 86 unchanged - 19 fixed = 103 total (was 105)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 37s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 43s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 34s |  the patch passed  |
| :green_heart: |  findbugs  |   4m  8s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 161m 34s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   | 218m 35s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux e02a6dfc3045 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 7ee6d59ef8 |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/3/artifact/out/diff-checkstyle-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/3/testReport/ |
| Max. process+thread count | 4915 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,288 @@
+    {
+      LOG.warn("regionCountCost is not set to 0, "
+              + " this will interfere with the HeterogeneousRegionCountCostFunction!");
+    }
```
Why is this a problem? Shouldn't we abort this cost function loading if this will cause it to behave improperly?

PierreZ replied:
`regionCountCost` is trying to put all RS to the same number of regions. I'm not sure that we should fire a Exception, and skip the init of the cost function

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,288 @@
+    final List<String> lines = HeterogeneousRegionCountCostFunction.readFile(this.rulesPath);
+    if (lines.size() == 0) {
+      return;
+    }
```
Is this cost function still capable of running if no rules are loaded? If not, shouldn't we throw an error to abort it's loading?

PierreZ replied:
Yes, default value will be used

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,288 @@
+      LOG.error("cannot read rules file located at ' " + filename + " ':" + e.getMessage());
+    }
+    return records;
+  }
```
This reads from HDFS only. Will read from local FS if HDFS is configured in local mode, instead of distributed, but that is only reasonable for tests purposes.

PierreZ replied:
In my opinion, this should work on both cases, depending on the prefix. The current implementation is working on local, not tested with HDFS

wchevreuil replied:
It will work when using hbase with local file system mode, but not when running in distributed mode. I just confirmed that by trying similar code on a cluster in distributed mode:

`WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
java.lang.IllegalArgumentException: Wrong FS: file:/root/test1.txt, expected: hdfs://mycluster.com:8020
	at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:662)`

Anyways, I think it's fine to leave it reading from HDFS only, we would just need to adjust the comments in this method and in the class description.

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,288 @@
+
+/**
+ * This is an optional Cost function designed to allow region count skew across RegionServers.
+ * A rule file is loaded from the local FS or HDFS before balancing. It contains lines of rules.
```
May need to review this assertion regarding local FS. See comment for _readFile_.

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,288 @@
+
+      // retrieve capacity for each RS
+      final ServerName sn = this.cluster.servers[i];
+      final double limit = this.findLimitForRS(sn);
```
nit: Could we just do _this.limitPerRS.get(serverName);_, then remove extra if block from lines 261 to 263? I guess at this point we would have a builtin _limitPerRS_ cache.

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,288 @@
+      final double limit = this.findLimitForRS(sn);
+      final double nbrRegions = this.cluster.regionsPerServer[i].length;
+
+      // cost is the number of regions above the local limit
```
nit: duplicate comment on line 149

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 29s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 55s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 32s |  master passed  |
| :green_heart: |  shadedjars  |   4m 57s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 39s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 30s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 28s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 27s |  the patch passed  |
| :green_heart: |  compile  |   0m 57s |  the patch passed  |
| :green_heart: |  javac  |   0m 57s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 28s |  hbase-server: The patch generated 71 new + 12 unchanged - 93 fixed = 83 total (was 105)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 56s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m  3s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 32s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 158m  9s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 26s |  The patch does not generate ASF License warnings.  |
|  |   | 219m 21s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.balancer.TestStochasticLoadBalancerHeterogeneousCost |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 603fe0d78fb4 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 2144690fb8 |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/4/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/4/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/4/testReport/ |
| Max. process+thread count | 4396 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/4/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

PierreZ commented:
Wow, so much checkstyles failed 😄 

Sorry for the disturbance @wchevreuil, I took into account most of your comments, thank you again 👍 

I will import checkstyles into Intellij to avoid pushing useless commits and fix it asap.

PierreZ commented:
Yes sorry about that auto-formatted thing. I will do that asap, thank you for the explanations @wchevreuil 😄

wchevreuil commented:
NP, @PierreZ . Also noticed TestStochasticLoadBalancerHeterogeneousCost is failing consistently for me as well. I guess some of the refactoring on last commit broke it?

PierreZ commented:
Thanks @wchevreuil, I updated the PR. Normally I should now have the right checkstyle in Intellij. I had another extension(Save actions) which created the mess :thinking:  

I added some testcases, I now need to work on the file part, to make it available from either HDFS or localFS. Thank you for your comments about this BTW.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m  6s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 23s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 20s |  master passed  |
| :green_heart: |  shadedjars  |   4m 38s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 37s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 14s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 12s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   4m 58s |  the patch passed  |
| :green_heart: |  compile  |   0m 57s |  the patch passed  |
| :green_heart: |  javac  |   0m 57s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 17s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 34s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 44s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m  5s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 340m 43s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 38s |  The patch does not generate ASF License warnings.  |
|  |   | 398m  6s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.master.TestSplitWALManager |
|   | hadoop.hbase.client.TestFromClientSide3 |
|   | hadoop.hbase.client.TestSnapshotTemporaryDirectory |
|   | hadoop.hbase.client.TestSnapshotDFSTemporaryDirectory |
|   | hadoop.hbase.util.TestFromClientSide3WoUnsafe |
|   | hadoop.hbase.client.TestAsyncTableAdminApi |
|   | hadoop.hbase.client.TestFromClientSide |
|   | hadoop.hbase.client.TestConnection |
|   | hadoop.hbase.client.TestFromClientSideWithCoprocessor |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 7a42bd15c872 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 7924ba39e7 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/5/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/5/testReport/ |
| Max. process+thread count | 4844 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/5/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

PierreZ commented:
I added a test to load the rule files from HDFS

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 55s |  master passed  |
| :green_heart: |  compile  |   0m 57s |  master passed  |
| :green_heart: |  checkstyle  |   1m 30s |  master passed  |
| :green_heart: |  shadedjars  |   4m 58s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 36s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 31s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 29s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 26s |  the patch passed  |
| :green_heart: |  compile  |   0m 56s |  the patch passed  |
| :green_heart: |  javac  |   0m 56s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 29s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 55s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m  9s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 16s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 156m 40s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 30s |  The patch does not generate ASF License warnings.  |
|  |   | 218m 34s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 1f969b0b19c3 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 395cfceb0b |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/6/testReport/ |
| Max. process+thread count | 4265 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/6/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

wchevreuil commented on hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java:
```diff
@@ -0,0 +1,306 @@
+    this.limitPerRS = new HashMap<>();
+    this.limitPerRule = new HashMap<>();
+    this.setMultiplier(conf.getFloat(REGION_COUNT_SKEW_COST_KEY, DEFAULT_REGION_COUNT_SKEW_COST));
+
```
Let's adhere to hbase's project code convention: 

> avoid lines with nothing but whitespace

See [here](https://hbase.apache.org/book.html#common.patch.feedback.trailingspaces).

wchevreuil replied:
I meant the empty lines within methods bodies should be avoided. For example, in this method, lines #103, #105, #113. In **cost()**  further below, lines #133, #135, #139, etc could all be ripped off.

PierreZ replied:
done :smile:

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 2 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 56s |  master passed  |
| :green_heart: |  compile  |   0m 57s |  master passed  |
| :green_heart: |  checkstyle  |   1m 31s |  master passed  |
| :green_heart: |  shadedjars  |   5m  0s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 37s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 30s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 27s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 31s |  the patch passed  |
| :green_heart: |  compile  |   0m 56s |  the patch passed  |
| :green_heart: |  javac  |   0m 56s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 29s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 56s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 12s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 41s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 154m 31s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   | 216m 17s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/677 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 42fec3a0d171 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-677/out/precommit/personality/provided.sh |
| git revision | master / 2ad62b0162 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/7/testReport/ |
| Max. process+thread count | 4811 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-677/7/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.
